### PR TITLE
Remove Subscribed from Frontend

### LINKED ISSUE #62: Subscribed Issues and Rendering
So I've been wracking my brain around the "three-pane toggle" bug that we identified at the end of last week. I have a fairly concrete idea of what the issue is (discussed below), but the several attempts at finding a reasonably simple solution have all failed. So I'm at a point where I think I should run it past you guys.

## The Problem ##  
First, recall what is happening. There is a new example, both on brax-dev and my recently pushed branch eric-subscribed-issues [under the resizable_panel demos](https://github.com/APrioriInvestments/object_database/blob/eric-subscribed-issues/object_database/web/cells_demo/resizable_panel.py#L63). You will see that if you toggle any one thing on then off, it all works fine. But if you toggle off the first pane, then either the second or third, then try to once more toggle (on) the first pane, the browser will error out.
  
At the Cells level, here is why that is happening. Look at this part of the code:
```python
return toggles() + cells.Flex(
            cells.Subscribed(
                lambda: cells.ResizablePanel(
                    cells.Subscribed(firstDisplay),
                    cells.Subscribed(toggledSecondAndThird),
                    ratio=0.20,
                )
                if ss.showFirst
                else cells.Subscribed(toggledSecondAndThird)
            )
        )
```
The structure of this description is that when all panes are "on," there should be a Subscribed wrapping a ResizablePanel that itself contains two Subscribeds, the second of which might itself be another nested ResizablePanel (for panes 2 and 3 if both are enabled). Depending on the configuration, a pane might be rendered inside of a Panel or a ResizablePanel. It can change, and for reasons I'll get to in a second that's important.
  
But just looking again at this snippet you can see a problem right away: inside of the outer Subscribed we will either render:
* A ResizablePanel that has two child Subscribeds
* A nested Subscribed that has either a ResizablePanel or just a Pane or nothing.
  
Once we have the ResizablePanel, and then it goes away and is replaced by two immediately nested Subscribeds, when we try to add it back in the frontend has no way of knowing where it should go anymore.
  
Here is how Subscribed works on the frontend. Any Subscribed component will do an initial `build()` on its content child, and then add the `data-subscribed-to=<Subscribed.id>` to the child content that actually gets rendered. Later, when its time to update that Subscribed, it looks up its last entrypoint in the DOM by querying for the element that has the matching `data-subscribed-to` data member.
  
Recall that in this example we first render a ResizablePanel, but then it goes out of the DOM. Now when trying to toggle it back on there is no element that has `data-subscribed-to=<Subscribed.id>` -- that would have been the ResizablePanel. 
  
## What I've Tried ##
I briefly tried reversing to the "old way" of simply rendering Subscribed components as plain `div` elements where needed in the rendering tree. This predictably didn't work -- it breaks styling and layout all over the place and leads to unpredictable visual results (this is why we stopped doing it in the first place).
  
I also attempted to rewire how the Subscribed component "knows" where to render its contents, primarily by laying down a placeholder for itself whenever the content is empty so it can come back and re-insert the real thing there. The problem with this methods occurs when you have a direct parent-child coupling of two Subscribed components (as we do in the offending example here). Making exceptions for that led to too much complexity and unexpected behavior.
  
## What Now ##
I am starting to run out of ideas, but I have one more that might work. I need your input on it. It hinges on the following question: Why do we need to send Subscribed over the wire at all and why does it need to be a Component? I think we should be able to determine on the Python side whether or not a Subscribed's content has changed (and therefore its parent description needs to be updated over the wire), rather than doing anything on the frontend. I think we already have a semi-workable framework making this possible, where the "true" parent is different from the "display" parent.
  
Do you think that could work? Am I missing something that might make this approach unfeasible?

### PR DESCRIPTION
**Pull request by darth-cheney**
This PR implements changes that ensure that the frontend no longer receives `Subscribed` cell descriptions.

## Motivation and Context

Issue #62 highlighted some deficiencies in how the frontend handles `Subscribed` Cells and their contents. What we realized in the course of working this out is that the frontend doesn't actually need to implement a `Subscribed` component in the first place: since all it does is receive updates from the structure on the Python side, we can "pretend" that the contents of any given `Subscribed` are the children of the given `Subscribed`'s parent -- ie make it "see through."

## Approach

The fix involved two major changes: lifecycle updating and Messenger changes.
  
### Lifecycle Updating ###
Normally, if the contents of a `Subscribed` change but the `Subscribed`'s parent or siblings haven't changed, it will be the "stable parent" for a given update message. But now that we are making these Cell types "see through," we need to ensure that when a `Subscribed` has changed during a given recalculation pass -- which means its `wasUpdated` property is `True` -- then we need to also force its parent to have `wasUpdated` equal `True`.
  
You can see this implemented in `_recalculateCells()` [here](https://github.com/APrioriInvestments/object_database/blob/eric-subscribed-issues/object_database/web/cells/cells.py#L516). For quick reference, the added code looks like this:
```python
        # Any Subscribeds that are set to broadcast
        # and that have been updated should also
        # make sure their parent Cell is updated.
        for node in self._nodesToBroadcast:
            if isinstance(node, Subscribed) and node.wasUpdated:
                node.parent.wasUpdated = True
```
The reason we cannot include this change inside of `recalculateSingleCell()` is because we are traversing by level, and doing the parent cells first. That means the parent of a Subscribed won't know if it's child has been updated until the next full pass has been done.
  
With this change, the `findStableParent()` method, called in `renderMessages()`, will now *never return a `Subscribed` cell*, but will instead return its parent or any ancestor that fits the stable criteria.
  
### Messenger Changes ###
The next set of changes that needed to be made were to ensure that when serializing Cell trees we "pass through" any encountered `Subscribed` Cells, and instead serialize their contents **as if those contents were the direct children of the Subscribed's parent**. By doing so, we never send any description of a `Subscribed` to the frontend, and the client (should) never even try to instantiate the now deprecated Subscribed component.
  
There are several places that we had to make this change, so feel free to browse the [Messenger.py file](https://github.com/APrioriInvestments/object_database/blob/eric-subscribed-issues/object_database/web/cells/Messenger.py).

## How Has This Been Tested?

We have added a new test file that directly tests for the new structure and pattern. All of those tests are currently passing.
  
Additionally, we manually tested the [CellsTestPage example for the three-button resizable pane toggling](https://github.com/APrioriInvestments/object_database/blob/eric-subscribed-issues/object_database/web/cells_demo/resizable_panel.py#L63). That is also working locally.

## Closes
#62 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: object_database/web/content/components/Subscribed.js [hunk 1]
```diff
@@ -21,6 +21,9 @@ class Subscribed extends Component {
 
         // Bind component methods
         this.makeContent = this.makeContent.bind(this);
+
+        // Debugging.
+        throw new Error('Subscribeds should not be created anymore!');
```
Comment (braxtonmckee) [inline object_database/web/content/components/Subscribed.js]: shouldn't we just get rid of this file?

File: object_database/web/cells/cells.py [hunk 2]
```diff
@@ -512,6 +513,13 @@ def _recalculateCells(self):
                     if not node.garbageCollected:
                         self._recalculateSingleCell(node)
 
+        # Any Subscribeds that are set to broadcast
+        # and that have been updated should also
+        # make sure their parent Cell is updated.
+        for node in self._nodesToBroadcast:
+            if isinstance(node, Subscribed) and node.wasUpdated:
+                node.parent.wasUpdated = True
```
Comment (braxtonmckee) [inline object_database/web/cells/cells.py]: what happens if you have a Subscribed in a Subscribed in a Subscribed? I think you probably need to walk up the tree until you see a node that's not a subscribed, marking wasUpdated the whole way.

Comment (darth-cheney) [reply object_database/web/cells/cells.py]: Good catch. Updated in the latest push.

File: object_database/web/content/components/Subscribed.js (hunk 1 above)
Comment (darth-cheney) [reply object_database/web/content/components/Subscribed.js]: I was thinking of "deprecating" it temporarily, but you're right -- we should remove it completely as well as any tests that rely on it. I've done so in the latest push. Tests are all passing.

**Pull request merged**